contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -1202,6 +1211,205 @@ CAPIStatus Filter::getStringFilterState(absl::string_view key, GoString* value_s
   return CAPIStatus::CAPIOK;
 }
 
+CAPIStatus Filter::getStringProperty(absl::string_view path, GoString* value_str, int* rc) {
+  // lock until this function return since it may running in a Go thread.
+  Thread::LockGuard lock(mutex_);
+  if (has_destroyed_) {
+    ENVOY_LOG(debug, "golang filter has been destroyed");
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  // to access the headers_ and its friends we need to hold the lock
+  activation_request_headers_ = dynamic_cast<const Http::RequestHeaderMap*>(request_headers_);
+  if (enter_encoding_) {
+    activation_response_headers_ = dynamic_cast<const Http::ResponseHeaderMap*>(headers_);
+    activation_response_trailers_ = dynamic_cast<const Http::ResponseTrailerMap*>(trailers_);
+  }
+
+  if (state.isThreadSafe()) {
+    const StreamInfo::StreamInfo& info = state.streamInfo();
```

Comment:
@wangfakang 
I have tried and it can't be done with a function pointer, as this part of code requires 'this' pointer, which is not available in non-member function.
Maybe we don't need the aggressive refactor, as other methods in this file also have repeated code in between `isThreadSafe` branch and not `isThreadSafe` branch.

Reply:
I have submitted a new commit to refactor.